### PR TITLE
Add TLS to datasource-syncer

### PR DESCRIPTION
**Pull request by macxamin**
1. Added TLS support, including four new flags `tls-cert`, `tls-key`, `tls-ca-cert` and `insecure-skip-verify`.
2. Added a condition to return error and exist if any data source syncs fail.
3. Added a label to datasource-syncer.yaml so users can use `kubectl get pods -l app=datasource-syncer` to view pods associated with their job.

### CODE REVIEW

File: cmd/datasource-syncer/main.go [hunk 1]
```diff
@@ -32,6 +34,11 @@ var (
 
 	gcmEndpointOverride = flag.String("gcm-endpoint-override", "",
 		"gcm-endpoint-override is the URL where queries should be sent to from Grafana. This should be left blank in almost all circumstances.")
+
+	certFile           = flag.String("tls-cert", "", "Path to the server TLS certificate.")
+	keyFile            = flag.String("tls-key", "", "Path to the server TLS key.")
+	caFile             = flag.String("ca-cert", "", "Path to the server certificate authority")
```
Comment (TheSpiritXIII) [inline cmd/datasource-syncer/main.go:40]: nit: name it `tls-ca-cert`?

File: cmd/datasource-syncer/main.go [hunk 2]
```diff
@@ -32,6 +34,11 @@ var (
...
+	InsecureSkipVerify = flag.Bool("insecure-skip-verify", false, "Skip TLS certificate verification")
```
Comment (TheSpiritXIII) [inline cmd/datasource-syncer/main.go:41]: nit: lowercase `insecureSkipVerify`?

File: cmd/datasource-syncer/main.go [hunk 3]
```diff
@@ -222,3 +238,43 @@ func buildUpdateDataSourceRequest(dataSource grafana.DataSource, token string) (
...
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		tlsConfig.RootCAs = caCertPool
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	client = &http.Client{Transport: transport}
```
Comment (TheSpiritXIII) [inline cmd/datasource-syncer/main.go:278]: suggestion: you should use the default client and update the transport.

File: cmd/datasource-syncer/main.go [hunk 4]
```diff
@@ -222,3 +238,43 @@ func buildUpdateDataSourceRequest(dataSource grafana.DataSource, token string) (
...
 	return &dataSource, nil
 }
+
+func getTLSClient(logger log.Logger, certFile, keyFile, caFile string, insecureSkipVerify bool) (*http.Client, error) {
+	if (certFile != "" || keyFile != "") && (certFile == "" || keyFile == "") {
+		return nil, fmt.Errorf("--tls-cert and tls-key must both be set or unset")
+	}
+
+	if certFile == "" && keyFile == "" && caFile == "" && !insecureSkipVerify {
+		return http.DefaultClient, nil
```
Comment (TheSpiritXIII) [inline cmd/datasource-syncer/main.go:248]: suggestion: the grafana library actually uses https://github.com/hashicorp/go-cleanhttp. Alternatively, just return nil and let the library decide the defaults.

File: cmd/datasource-syncer/main.go [hunk 5]
```diff
@@ -222,3 +238,43 @@ func buildUpdateDataSourceRequest(dataSource grafana.DataSource, token string) (
...
+		caCert, err := os.ReadFile(caFile)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read ca cert: %w", err)
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		tlsConfig.RootCAs = caCertPool
+	}
+
+	transport := &http.Transport{
```
Comment (TheSpiritXIII) [inline cmd/datasource-syncer/main.go:274]: suggestion: use default transport and update its values.

**Review by TheSpiritXIII — CHANGES_REQUESTED** [review]
Looks good! Just left some minor nits and suggestions.

File: cmd/datasource-syncer/main.go (hunk 5 above)
Comment (macxamin) [reply cmd/datasource-syncer/main.go:273]: breaks code.

File: cmd/datasource-syncer/main.go (hunk 3 above)
Comment (macxamin) [reply cmd/datasource-syncer/main.go]: breaks code.

File: cmd/datasource-syncer/main.go (hunk 5 above)
Comment (TheSpiritXIII) [reply cmd/datasource-syncer/main.go:281]: Can you put this in a comment so we understand why we didn't use it?

**Review by TheSpiritXIII — APPROVED** [review]
Thank you!

**Pull request merged**